File: CPP/ByRound/1693/1693_C_Keshi_in_Search_of_AmShZ.cpp
// Problem: CF 1693 C - Keshi in Search of AmShZ
// https://codeforces.com/contest/1693/problem/C

/*
 * Problem: Keshi in Search of AmShZ
 *
 * Approach:
 * This problem is about finding the minimum number of days required for Keshi
 * to reach city n from city 1, where each day AmShZ can either block a road or
 * tell Keshi to move. Keshi moves to a random reachable city (if any), and
 * AmShZ knows Keshi's location at all times.
 *
 * Key Idea:
 * This is essentially a shortest path problem, but we're looking for a
 * "worst-case" scenario. We model it as a modified Dijkstra using a priority
 * queue.
 *
 * We calculate d[i] = minimum number of days to reach city i from city n,
 * and for each city we track how many outgoing edges remain unblocked.
 *
 * We use a reverse graph (rg) to process from the destination (n) back to
 * source (1), using a modified Dijkstra with priority on the number of
 * out-going edges.
 *
 * Time Complexity: O((n + m) * log n)
 * Space Complexity: O(n + m)
 */

#include <bits/stdc++.h>
using namespace std;

const int N = 2e5 + 5;
int n, m, x, y, out[N], d[N], v[N];
vector<int> rg[N]; // Reverse graph: rg[i] contains cities with edges to i
priority_queue<pair<int, int>>
    q; // Priority queue for Dijkstra: {-distance, node}

signed main() {
  scanf("%d%d", &n, &m);
  // Read all directed edges and update out-degree and reverse graph
  for (int i = 1; i <= m; i++)
    scanf("%d%d", &x, &y), out[x]++, rg[y].push_back(x);

  // Initialize distances to infinity (0x3f = 63 in decimal)
  memset(d, 0x3f, sizeof(d)),
      d[n] = 0, q.push({0, n}); // Start from node n with distance 0
  while (q.size()) {
    int y = q.top().second; // Current node
    q.pop();
    if (v[y]) // Skip if already processed (visited)
      continue;
    v[y] = 1; // Mark as visited

    // For all predecessors of current node y
    for (int x : rg[y]) {
      // If we found a better way to reach x, update distance
      if (d[x] > d[y] + out[x])
        d[x] = d[y] + out[x], q.push({-d[x], x});
      out[x]--; // Decrease the out-degree of x as one edge is blocked
    }
  }
  printf("%d\n", d[1]); // Distance from city 1 to city n
  return 0;
}

// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/C/1693_C_Keshi_in_Search_of_AmShZ.py
# Problem: CF 1693 C - Keshi in Search of AmShZ
# https://codeforces.com/contest/1693/problem/C

"""
Purpose: 
Find the minimum number of days required for Keshi to reach city n from city 1,
given that AmShZ can block roads or instruct Keshi to move each day.
This problem is modeled as a shortest path problem on a directed graph where:
- Each node represents a city.
- Each edge represents a directed road.
- We use a modified Dijkstra's algorithm with a priority queue to find the minimum steps.
- For each node, we track the minimum number of operations needed to reach it.
- The weight of moving from a node is based on the in-degree of the neighbor node.
- This is essentially a reverse topological traversal with weighted edges, using Dijkstra's.

Time Complexity: O((V + E) * log V), where V is number of cities and E is number of roads.
Space Complexity: O(V + E), for storing the graph and auxiliary arrays.
"""

from sys import stdin

input = lambda: stdin.readline()[:-1]

n, m = map(int, input().split())
edge = [[] for i in range(n)]
ord = [0] * n
for _ in range(m):
    a, b = map(lambda x: int(x) - 1, input().split())
    edge[b].append(a)
    ord[a] += 1

inf = 1 << 30
dist = [inf] * n
from heapq import heappop, heappush

# Start from city n (destination) and compute distances backward
hq = [(0, n - 1)]
dist[n - 1] = 0
seen = [False] * n
while hq:
    w, v = heappop(hq)
    if dist[v] < w:
        continue
    seen[v] = True
    # Traverse all predecessors of v
    for to in edge[v]:
        # Update the distance if going through this edge is better
        if dist[v] + ord[to] < dist[to]:
            dist[to] = dist[v] + ord[to]
            heappush(hq, (dist[to], to))
        # Decrease the out-degree of the predecessor (simulate blocking)
        ord[to] -= 1

# The answer is the minimum distance to reach city 1 (index 0)
print(dist[0])


# https://github.com/VaHiX/CodeForces/